File: LeetCode-DSA/DataStructures-1/PascalTriangle.cpp
/*
Q118. Description: Given an integer numRows, return the first numRows of Pascal's triangle.
In Pascal's triangle, each number is the sum of the two numbers directly above it as shown
*/

class Solution {
public:
    int binomial_coef(int n, int k)
    {
        // Find binomial coefficient of every term in (x+y)^n nCk
        int res = 1;
        if(k > n)
            k = n-k;
        for(int i = 0; i < k; i++)
        {
            res = res * (n-i);
            res = res / (i+1);
        }
        return res;
    }
    
    vector<vector<int>> generate(int numRows) 
    {
        vector<vector<int>> result;
        for(int i = 0; i < numRows; i++)
        {
            // For every row, find coefficient for k = 0,1...row number
            vector<int> temp;
            for(int j = 0; j <= i; j++)
                temp.push_back(binomial_coef(i,j));
            result.push_back(temp);
        }
        return result;
    }
};

File: LeetCode-DSA/JulyChallenge_2022/PascalTriangle.py
'''
Q118. Description: Given an integer numRows, return the first numRows of Pascal's triangle.
In Pascal's triangle, each number is the sum of the two numbers directly above it.
'''

class Solution:
    def coef(self, n, k):
        # Find binomial coefficient of current row and element
        res = 1
        if (k > n-k):
            k = n-k
        for i in range(0, k):
            res = res * (n-i)
            res = res // (i+1)
        return res
    
    def generate(self, numRows: int) -> List[List[int]]:
        sol = []
        for line in range(numRows):
            curr = []
            for i in range(0, line+1):
                curr.append(self.coef(line, i))
            sol.append(curr)
        return sol
        